raiden/network/rpc/client.py

```diff
@@ -686,7 +686,17 @@ def eth_estimateGas(
             startgas,
             gasprice,
         )
-        res = self.call('eth_estimateGas', json_data)
+        try:
+            res = self.call('eth_estimateGas', json_data)
+        except EthNodeCommunicationError as e:
+            tx_would_fail = str(e) in [
+                'gas required exceeds allowance or always failing transaction',  # geth
+                'Transaction execution error.'  # parity
+            ]
+            if tx_would_fail:
+                return -1
```

Comment:
Returning an invalid gas amount doesn't seem like a good idea.

Alternatives:
1. Return another type (None)
1. Return a constant error object
1. Raise a Raiden exception

I think the 3. option is the best one in this case, since an execution error in the smart contract is an unrecoverable scenario.

Reply:
It depends on how we use estimate gas. It can premptively detect a transaction that would throw so we have to see what kind of pattern we want to use. Either a return value or an exception. 

I need to think about this a bit more. I made also another issue on the usage of this estimate gas function [here](https://github.com/raiden-network/raiden/issues/1216).

Reply:
> It can premptively detect a transaction that would throw so we have to see what kind of pattern we want to use

What do you mean by "preemptively"?

Reply:
> Either a return value or an exception.

Anyways, I think that `-1` is a bad choice.

Returning a value should be used for functions that have failing behavior, e.g. a call to `transfer` or the call to tinyrpc itself (which should not throw, but oh well). Nevertheless the returned value should be an object of a different type, were the type tells you that an error happened (because python doesn't have an Option/Maybe/Either types I proposed options 1 and 2).

Raising an exception should be used on exceptional/unrecoverable cases, and although I'm not 100% sure we should treat failures to smart contracts as exceptional, I don't think raising an exception from the estimate function will result in exception for control flow, so that may turn out okay.

Reply:
yes `None` is definitely a better choice. I will make this change.

I just don't feel right throwing an exception for estimate gas finding out a tx would throw as this may be recoverable, but I made a new issue just so we can think a bit further on this matter.